lib/snappy_ex.ex

```diff
@@ -9,55 +9,87 @@ defmodule LambdaEthereumConsensus.SnappyEx do
     end
   end
 
+  defp decompress_frames(chunks, acc) when byte_size(chunks) == 0, do: acc
+
   defp decompress_frames(chunks, acc) do
-    if byte_size(chunks) != 0 do
-      {chunk_id, chunks} = process_chunk_id(chunks)
-      {chunks_size, chunks} = process_chunk_size(chunks)
-      {acc, chunks} = process_chunks(chunks, acc, chunks_size, chunk_id)
+    with {:ok, {chunk_id, chunk_size, chunks}} <- process_chunk_metadata(chunks),
+         {:ok, {acc, chunks}} <- process_chunks(chunks, acc, chunk_size, chunk_id) do
       decompress_frames(chunks, acc)
-    else
-      acc
     end
   end
 
-  defp process_chunk_id(<<chunk_id::size(8), chunks::binary>>), do: {chunk_id, chunks}
+  # chunk layout: 1-byte chunk_id, 3-bytes chunk_size, remaining chunks if any data present.
+  defp process_chunk_metadata(chunks) when byte_size(chunks) < 4,
+    do: {:error, "invalid chunks metadata size"}
 
-  defp process_chunk_size(<<chunk_size::little-integer-size(24), chunks::binary>>),
-    do: {chunk_size, chunks}
+  defp process_chunk_metadata(chunks) do
+    <<chunk_id::size(8), chunks_size::size(24), chunks::binary>> = chunks
```

Comment:
You have to specify little-endian for the `chunks_size` (the default is big-endian, and the spec uses little):
```suggestion
    <<chunk_id, chunks_size::little-size(24), chunks::binary>> = chunks
```